rust/exchange_rate/src/main.rs

```diff
@@ -0,0 +1,385 @@
+    pub url: String,
+    pub max_response_bytes: Option<u64>,
+    pub headers: Vec<HttpHeader>,
+    pub body: Option<Vec<u8>>,
+    pub http_method: HttpMethod,
+    pub transform_method_name: Option<String>,
+}
+
+#[derive(CandidType, Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
+pub struct CanisterHttpResponsePayload {
+    pub status: u64,
+    pub headers: Vec<HttpHeader>,
+    pub body: Vec<u8>,
+}
+
+// How many data point can be returned as maximum.
+// Given that 2MB is max-allow cansiter response size, and each <Timestamp, Rate> pair
+// should be less that 20 bytes. Maximum data points could be returned for each
+// call can be as many as 2MB / 20B = 1000000.
+pub const MAX_DATA_PONTS_CANISTER_RESPONSE: usize = 1000000;
+
+// Remote fetch interval in secs. It is only the canister returned interval
+// that is dynamic according to the data size needs to be returned.
+pub const REMOTE_FETCH_GRANULARITY: u64 = 60;
+
+// For how many rounds of heartbeat, make a http_request call.
+pub const RATE_LIMIT_FACTOR: usize = 5;
+
+// How many data points in each Coinbase API call. Maximum allowed is 300
+pub const DATA_POINTS_PER_API: u64 = 200;
+
+// Maximum raw Coinbase API response size. This field is used by IC to calculate circles cost per HTTP call.
+// Here is how this number is derived:
+// Each Coinbase API call return an array of array, and each sub-array look like below:
+// [
+//     1652454000,
+//     9.51,
+//     9.6,
+//     9.55,
+//     9.54,
+//     4857.1892
+// ],
+// Each field of this sub-arry takes less than 10 bytes. Then,
+// 10 (bytes per field) * 6 (fields per timestamp) * 200 (timestamps)
+pub const MAX_RESPONSE_BYTES: u64 = 10 * 6 * DATA_POINTS_PER_API;
+
+pub const RESPONSE_HEADERS_SANTIZATION: [&'static str; 7] = [
+    "Date",                    // DateTime of the request is made
+    "CF-Cache-Status",         // CloudFront caching status
+    "CF-RAY",                  // CloudFront custom Id
+    "Age",                     // Age of the data object since query
+    "Content-Security-Policy", // Long list of allowable domains for reference
+    "Last-Modified",           // Last time the object is modified
+    "Set-Cookie",              // cf-country=US;Path=/;
+];
+
+thread_local! {
+    pub static FETCHED: RefCell<HashMap<Timestamp, Rate>>  = RefCell::new(HashMap::new());
+    pub static REQUESTED: RefCell<HashSet<Timestamp>> = RefCell::new(HashSet::new());
+    pub static RATE_COUNTER: RefCell<usize> = RefCell::new(0);
+}
+
+// Canister heartbeat. Process one item in queue
+// #[export_name = "canister_heartbeat"]
+#[heartbeat]
+async fn heartbeat() {
+    let mut should_fetch = false;
+    RATE_COUNTER.with(|counter| {
+        let state = counter.clone().into_inner();
+        if state == 0 {
+            should_fetch = true;
+        }
+        counter.replace((state + 1) % RATE_LIMIT_FACTOR);
+    });
+    if should_fetch {
+        get_next_rate().await;
```

Comment:
Awaiting a function within heartbeat is already a relatively risky thing, and this function can even make even more external calls. Can you add some comment about cycles use and/or canister safety here?